### PR TITLE
Convert more things to components in TS

### PR DESCRIPTION
**Pull request by jrsquared**
Convert accountLabelColor, execution, executionDetails, executionService, schedulerFactory, executionFilter.model, executionFilter.service to TS.

For executionDetails, I only converted the directive. The spaghetti mess of executionDetails controllers is well beyond the scope of this conversion, and worth a revisit later. 

One caveat to this is in the component constructor, I just put the controller bindings back onto the scope to obviate the need to modify every details controller and template. I made a TODO comment along these lines.

I also changed `execution.showingDetails` from a function to a boolean. The boolean gets updated as appropriate state changes. (Removes the need to regex the state every time showingDetails gets checked and only runs the regex when state changes)

### CODE REVIEW

File: app/scripts/modules/core/delivery/executionGroup/execution/execution.component.ts [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+  private updateViewStateDetails(): void {
+    this.viewState.activeStageId = Number(this.$stateParams.stage);
+    this.viewState.executionId = this.$stateParams.executionId;
+    this.invalidateShowingDetails();
+  };
+
+  private invalidateShowingDetails(): void {
+    // DONE: watch standalone and execution in $onChanges()...
+    // DONE: Make this a boolean instead of a function, then these changes will modify the boolean
+    // DONE: Actually use showingDetails instead of showDetails()
```
Comment (anotherchrisberry) [inline app/scripts/modules/core/delivery/executionGroup/execution/execution.component.ts]: Great comments

File: app/scripts/modules/core/cancelModal/cancelModal.service.ts [hunk 2]
```diff
@@ -8,7 +8,7 @@ export interface ICancelModalParams {
   cancelButtonText?: string;
   forceable?: boolean;
   header: string;
-  submitMethod: (args: any) => ng.IPromise<any>;
+  submitMethod: (...args: any[]) => ng.IPromise<any>;
```
Comment (anotherchrisberry) [inline app/scripts/modules/core/cancelModal/cancelModal.service.ts]: Can we change this to match the expected arguments (string, optional boolean)?

File: app/scripts/modules/core/delivery/executionGroup/execution/execution.component.html [hunk 3]
```diff
@@ -0,0 +1,115 @@
+<div class="execution" id="execution-{{::$ctrl.execution.id}}" ng-class="$ctrl.showingDetails ? 'show-details' : 'details-hidden'">
+  <div class="execution-overview group-by-{{::$ctrl.sortFilter.groupBy}}">
+    <h4 class="execution-name" ng-if="::$ctrl.sortFilter.groupBy !== 'name'">
+      <account-label-color ng-repeat="account in $ctrl.execution.deploymentTargets" account="{{account}}"></account-label-color>
+      {{::$ctrl.execution.name}}
+    </h4>
+    <execution-status execution="::$ctrl.execution"
+                      toggle-details="$ctrl.toggleDetails(node)"
```
Comment (anotherchrisberry) [inline app/scripts/modules/core/delivery/executionGroup/execution/execution.component.html]: I don't think you have a `node` here.

File: app/scripts/modules/core/delivery/executionGroup/execution/execution.component.html [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+           analytics-on="click"
+           analytics-category="Pipeline"
+           analytics-event="Execution source clicked"
+           ng-href="{{$ctrl.pipelinesUrl + $ctrl.execution.id}}">Source</a>
+        |
+        <a ng-href="{{::$ctrl.getUrl()}}"
+           analytics-on="click"
+           analytics-category="Pipeline"
+           analytics-event="Permalink clicked">Permalink</a>
+        <copy-to-clipboard text="{{::$ctrl.getUrl()}}" tool-tip="Copy permalink to clipboard"></copy-to-clipboard>
```
Comment (anotherchrisberry) [inline app/scripts/modules/core/delivery/executionGroup/execution/execution.component.html]: We shouldn't one-time bind `$ctrl.getUrl()` (this is an existing bug, not something you introduced) - same as above on the `ng-href`

File: app/scripts/modules/core/delivery/executionGroup/execution/execution.component.html (hunk 3 above)
Comment (jrsquared) [inline app/scripts/modules/core/delivery/executionGroup/execution/execution.component.html]: Angular components pass properties via a map when bound with an `&`. Like https://docs.angularjs.org/guide/component. It's confusing.

Comment (anotherchrisberry) [inline app/scripts/modules/core/delivery/executionGroup/execution/execution.component.html]: 👍

File: app/scripts/modules/core/delivery/executionGroup/execution/execution.component.ts [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+    if (changesObj.standalone || changesObj.execution) {
+      this.invalidateShowingDetails();
+    }
+  }
+
+  public isActive(stageIndex: number): boolean {
+    return this.showingDetails && Number(this.$stateParams.stage) === stageIndex;
+  }
+
+  public toggleDetails(node: any): void {
```
Comment (anotherchrisberry) [inline app/scripts/modules/core/delivery/executionGroup/execution/execution.component.ts]: It would be nice to make this an inline interface, i.e.
```
public toggleDetails(node: {executionId: string, index: number}): void {
```

File: app/scripts/modules/core/delivery/status/executionStatus.component.ts [hunk 6]
```diff
@@ -8,8 +8,8 @@ import './executionStatus.less';
 
 export class ExecutionStatusController implements IComponentController {
   public execution: IExecution;
-  public toggleDetails: () => void;
-  public showingDetails: () => boolean;
+  public toggleDetails: (node: any) => void;
```
Comment (anotherchrisberry) [inline app/scripts/modules/core/delivery/status/executionStatus.component.ts]: super super cool if this can be an interface (either inlined or explicitly declared/exported somewhere)

File: app/scripts/modules/core/delivery/executionGroup/execution/execution.component.ts (hunk 5 above)
Comment (jrsquared) [inline app/scripts/modules/core/delivery/executionGroup/execution/execution.component.ts]: GOOD call

File: app/scripts/modules/core/delivery/details/executionDetails.controller.js [hunk 7]
```diff
@@ -9,10 +9,18 @@ module.exports = angular.module('spinnaker.executionDetails.controller', [
   .controller('executionDetails', function($scope, $stateParams, $state, pipelineConfig) {
     var controller = this;
 
-    controller.standalone = $scope.standalone || false;
+    controller.standalone = controller.standalone || false;
+
+    // This is pretty dirty but executionDetails has it's dirty tentacles
```
Comment (anotherchrisberry) [inline app/scripts/modules/core/delivery/details/executionDetails.controller.js]: `its` - `it's` means `it is`

File: app/scripts/modules/core/delivery/details/executionDetails.controller.js [hunk 8]
```diff
@@ -9,10 +9,18 @@ module.exports = angular.module('spinnaker.executionDetails.controller', [
...
+    // all over the place. This makes the conversion of the execution direcive
```
Comment (anotherchrisberry) [inline app/scripts/modules/core/delivery/details/executionDetails.controller.js]: not how we spell `directive`

File: app/scripts/modules/core/delivery/details/executionDetails.controller.js [hunk 9]
```diff
@@ -9,10 +9,18 @@ module.exports = angular.module('spinnaker.executionDetails.controller', [
...
+    // to a component safe until we tackle converting all the controllers
+    // TODO: Convert all the exeuction details controllers to ES6 controllers and remove references to $scope
```
Comment (anotherchrisberry) [inline app/scripts/modules/core/delivery/details/executionDetails.controller.js]: or how we spell `execution`

**Comment by anotherchrisberry** [comment]
pulled this down and ran through it - everything seems fine, so nice job on that; will look a little closer at the code itself tomorrow.

File: app/scripts/modules/core/delivery/executionGroup/execution/execution.component.html [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+    </execution-status>
+    <div class="execution-bar">
+      <div class="stages" ng-class="{'show-durations': $ctrl.sortFilter.showStageDuration}">
+        <div ng-repeat="stage in ::$ctrl.execution.stageSummaries"
+             class="clickable stage execution-marker"
+             ng-style="::{width: 100/$ctrl.execution.stageSummaries.length + '%', 'background-color': stage.color}"
+             ng-class="['stage-type-' + stage.type.toLowerCase(), 'execution-marker-' + stage.status.toLowerCase(), $ctrl.isActive($index) ? 'active' : '', stage.isRunning ? 'glowing' : '']"
+             uib-tooltip-template="::stage.labelTemplateUrl"
+             tooltip-title="(Click for details)"
+             tooltip-placement="bottom"
```
Comment (anotherchrisberry) [inline app/scripts/modules/core/delivery/executionGroup/execution/execution.component.html]: while you're in here, might as well remove the `tooltip-title` and `tooltip-placement` attributes, since they're not used or doing anything

File: app/scripts/modules/core/delivery/executionGroup/execution/execution.component.ts [hunk 11]
```diff
@@ -0,0 +1,206 @@
...
+
+  static get $inject(): string[] { return ['$scope', '$location', '$stateParams', '$state', 'schedulerFactory',
+                                           'executionFilterModel', 'executionService', 'cancelModalService',
+                                           'confirmationModalService']; }
+
+  constructor(private $scope: IScope,
+              private $location: ILocationService,
+              private $stateParams: IExecutionDetailsStateParams,
+              private $state: IStateService,
+              private schedulerFactory: any,
```
Comment (anotherchrisberry) [inline app/scripts/modules/core/delivery/executionGroup/execution/execution.component.ts]: Can strongly type `schedulerFactory` now

File: app/scripts/modules/core/delivery/filter/executionFilter.service.spec.ts [hunk 12]
```diff
@@ -41,7 +45,7 @@ describe('Service: executionFilterService', function () {
   describe('Updating execution groups', function () {
 
     it('limits executions per pipeline', function () {
-      let application = {
+      let application: any = {
         executions: { data: [
           { pipelineConfigId: '1', name: 'pipeline 1', endTime: 1, stages: [] },
           { pipelineConfigId: '1', name: 'pipeline 1', endTime: 2, stages: [] },
```
Comment (anotherchrisberry) [inline app/scripts/modules/core/delivery/filter/executionFilter.service.spec.ts]: If you want to get truly TS-based, change this to:
```typescript
let application: Application = applicationModelBuilder.createApplication({key: 'executions', lazy: true});
application.getDataSource('executions').data = [ ... ];
```
but what you have works

File: app/scripts/modules/core/delivery/filter/executionFilter.service.ts [hunk 13]
```diff
@@ -0,0 +1,320 @@
...
+  // this gets called every time the URL changes, so we debounce it a tiny bit
+  public updateExecutionGroups: (application: Application) => IGroup[];
+
+
+
+  static get $inject(): string[] { return ['executionFilterModel', 'timeBoundaries', 'waypointService', '$log', 'filterModelService', 'pipelineConfig']; }
+
+  constructor(private executionFilterModel: ExecutionFilterModel,
+              private timeBoundaries: any,
+              waypointService: any,
```
Comment (anotherchrisberry) [inline app/scripts/modules/core/delivery/filter/executionFilter.service.ts]: Why no access modifier on `waypointService`? It doesn't need to be public, does it?

File: app/scripts/modules/core/delivery/filter/executionFilter.service.ts [hunk 14]
```diff
@@ -0,0 +1,320 @@
...
+        }
+      }
+      const executions = application.executions.data || [];
+      executions.forEach((execution: IExecution) => this.fixName(execution, application));
+      const filtered: IExecution[] = this.filterExecutionsForDisplay(application.executions.data);
+
+      const groups = this.groupExecutions(filtered, application);
+
+      this.applyGroupsToModel(groups);
+      waypointService.restoreToWaypoint(application.name);
```
Comment (anotherchrisberry) [inline app/scripts/modules/core/delivery/filter/executionFilter.service.ts]: ah, is this why there's no access modifier? it's only used here?

File: app/scripts/modules/core/delivery/service/execution.service.ts [hunk 15]
```diff
@@ -0,0 +1,358 @@
...
+    private waitUntilPipelineIsCancelled(application: Application, executionId: string): IPromise<any> {
+      return this.waitUntilExecutionMatches(executionId, (execution: IExecution) => execution.status === 'CANCELED')
+        .then(() => application.executions.refresh());
+    }
+
+    private waitUntilPipelineIsDeleted(application: Application, executionId: string): IPromise<any> {
+      const deferred = this.$q.defer();
+      this.getExecution(executionId).then(
+        () => this.$timeout(() => this.waitUntilPipelineIsDeleted(application, executionId).then(deferred.resolve), 1000),
+        deferred.resolve
```
Comment (anotherchrisberry) [inline app/scripts/modules/core/delivery/service/execution.service.ts]: any weird `this` binding behavior here? i.e. does it need to be `() => deferred.resolve()`?

**Comment by anotherchrisberry** [comment]
A few little things, LGTM otherwise (I may have glossed over the last 1500 lines)

File: app/scripts/modules/core/scheduler/scheduler.factory.ts [hunk 16]
```diff
@@ -0,0 +1,101 @@
...
+          scheduler.unsubscribe();
+        }
+        scheduler = null;
+        source = null;
+        this.$timeout.cancel(pendingRun);
+        document.removeEventListener('visibilitychange', watchDocumentVisibility);
+        this.$window.removeEventListener('offline', suspendScheduler);
+        this.$window.removeEventListener('online', resumeScheduler);
+      }
+    };
```
Comment (anotherchrisberry) [inline app/scripts/modules/core/scheduler/scheduler.factory.ts]: should make this into a distinct class with constructor

File: app/scripts/modules/core/delivery/filter/executionFilter.service.ts (hunk 14 above)
Comment (jrsquared) [inline app/scripts/modules/core/delivery/filter/executionFilter.service.ts]: Yeah, only used in the constructor.

File: app/scripts/modules/core/scheduler/scheduler.factory.ts (hunk 16 above)
Comment (jrsquared) [inline app/scripts/modules/core/scheduler/scheduler.factory.ts]: Agreed. Will do this as a separate PR.
